pallets/xcm-transactor/src/lib.rs

```diff
@@ -743,9 +746,19 @@ pub mod pallet {
 
 		/// Returns the fee for a given set of parameters
 		pub fn calculate_fee_per_second(weight: Weight, fee_per_second: u128) -> u128 {
-			let weight_fee =
-				fee_per_second.saturating_mul(weight as u128) / (WEIGHT_PER_SECOND as u128);
-			return weight_fee;
+			// We use perbill to be more precise over floating point mults
+			// floating point multiplications are rounded up/down depending on the fractional part
+			// for integer multiplication and then division, we would always round down
+			let fee_per_bill = Perbill::from_rational(
+				fee_per_second % (WEIGHT_PER_SECOND as u128),
+				WEIGHT_PER_SECOND as u128,
+			);
+			let weight_fee_frac = fee_per_bill * (weight as u128);
```

Comment:
To be clear, I was referring to `from_rational`, which is relatively expensive: https://github.com/paritytech/substrate/blob/19a8a9ba9fec2383c75cc5a1dd51d45d28c1f24b/primitives/arithmetic/src/per_things.rs#L466

I believe this will also truncate, in this case to 1 billion (a small number in the world of weight).

Reply:
Sorry the computation we need to do is `fee_per_second * weight / Weight_Per_Second`. So the solution of adding `Weight_Per_Second` before the division is a different operation.

I think the only way to round up `**always**` is doing

let temp  =  fee_per_second.saturating_mul(weight as u128)
`rust
let weight_fee = temp.saturating_add(WEIGHT_PER_SECOND.saturating_sub((temp%WEIGHT_PER_SECOND)))/WEIGHT_PER_SECOND`

Reply:
Something like this (this passes the tests BTW)

```rust
let temp  =  fee_per_second.saturating_mul(weight as u128);

let weight_per_second_u128 = WEIGHT_PER_SECOND as u128;
if temp%weight_per_second_u128 == 0 {
	return temp/weight_per_second_u128;
}
else {
	let weight_fee = 
temp.saturating_add(weight_per_second_u128.saturating_sub(temp%weight_per_second_u128))/weight_per_second_u128;
        return weight_fee
}
```

Reply:
makes sense, then lets add your idea. We add WEIGHT_PER_SECOND-1 to make sure we always round up in the computations.